keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,776 @@
+A user wants to create a cluster with 32 nodes each with a capacity to run 256 pods. This means that each node needs a /24 pod CIDR range and they need a total range of /19. However, there aren't enough contiguous IPs in the user's network. They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each with a /21 range. The CIDR allocator allocates a /24 range from any of these /21 ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+A major precondition for this feature is to prevent churn on the IPs assigned to Nodes. Specifically, we want to prevent the controller from attempting to change a Node’s assigned PodCIDRs. For example, K8s will reject attempts to delete a CIDRRange that’s currently in use by a Node. This should prevent accidental broad-ranging changes to the cluster.
+
+### Risks and Mitigations
+
+- Do racing controllers need a more sophisticated design? The current plan is to plug in to the kube-controller-manager leader election.
+- Can the controller install admission controls? This is a new pattern for in-tree components.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New Resource
+```
+ClusterCIDRConfig {
+    # Nodes may only have 1 range from each family.
+    # TODO: maybe this is redundant and can just be inferred by the controller.
+    IPFamily        FamilyType
+    
+    # An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    # +required
+    IPCIDRBlock     string
+    
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule 
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
+}
+
+var (
+    IPV4 FamilyType = "IPV4"
+    IPV6 FamilyType = "IPV6"
+)
+```
+
+#### Expected Behavior
+
+- Each node will be assigned up to one range from each `FamilyType`.
+    - **TODO:** what is the tie-break if multiple ranges match
+- An empty `NodeSelector` functions as a default that applies to all nodes. This should be the fall-back and not take precedence if any other range matches.
+- `IPFamily`, `IPCIDRBlock`, and `PerNodeMaskSize` are immutable after creation.
+- Reject any changes that would change existing node allocations:
+    - Can not delete a `CIDRRangeConfig` that is in use by a node
```

Comment:
how do you know if it is in use?
do we use to the PV/PVC model?

Reply:
We didn't have a good understanding of how validations would work. We don't have an idea of how we would do this.

Reply:
I think there's 2 approaches we could take here.

1. Use a admission controller that validates any changes before they're committed. So in this case, the request to delete the CIDRRangeConfig would come to the controller. The controller knows if a range is in use, so it could reject the request.
2. Use the PV/PVC model of putting the object in a "terminating" state and delete it when the last node using it goes away. We can probably attach a condition to the object listing the nodes still using the config.

I'm leaning towards option 2. My thinking is that this is a model that we've used before so should be familiar. Are there any major roadblocks you can think of if we were to try option 2.